### PR TITLE
Add RepLLaMA style models

### PR DESCRIPTION
**Pull request by orionw**
Adds [RepLLaMA models](https://arxiv.org/abs/2310.08319) and [Promptriever (paper on arxiv tomorrow)](https://huggingface.co/collections/samaya-ai/promptriever-66e068770af4dd364c9d3c7b) models as MTEB models. Verified they reproduced. Subtly, the training code adds an extra space as the prefix that I figured out the hard way several months ago.

## Checklist

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 

### Adding a model checklist

 - [x] I have filled out the ModelMeta object to the extent possible
 - [x] I have ensured that my model can be loaded using
   - [x] `mteb.get_model(model_name, revision_id)` and
   - [] `mteb.get_model_meta(model_name, revision_id)`
 - [x] I have tested the implementation works on a representative set of tasks.
 
 mteb.get_model_meta doesn't seem to work, it gives me
 
 ```
 AttributeError: 'ModelMeta' object has no attribute 'encode'
```

But perhaps I'm doing something wrong @KennethEnevoldsen? Do I use it in the same way as the second line below (e.g. `model = mteb.get_model_meta("samaya-ai/promptriever-llama2-7b-v1")`? Or perhaps it's because it's a peft model?

Also, weirdly it runs the encoding twice when I use this:
```
import mteb
model = mteb.get_model("samaya-ai/RepLLaMA-reproduced")
tasks = mteb.get_tasks(tasks=["SciFact"], languages=["eng"])
evaluation = mteb.MTEB(tasks=tasks)
evaluation.run(model, batch_size=16)
```

### CODE REVIEW

File: mteb/models/promptriever_models.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        RepLLaMAWrapper,
+        base_model_name_or_path="meta-llama/Llama-2-7b-hf",
+        peft_model_name_or_path="samaya-ai/promptriever-llama2-7b-v1",
+        device_map="auto",
+        torch_dtype=torch.bfloat16,
+    ),
+    name="samaya-ai/promptriever-llama2-7b-v1",
+    languages=["eng_Latn"],
+    open_source=True,
+    revision=None,  # TODO: Not sure what to put here as a model is made of two peft repos, each with a different revision
```
Comment (KennethEnevoldsen) [inline mteb/models/promptriever_models.py:58]: A quick fix is "rev1-rev2" as a joint revision for the joint model

**Review by KennethEnevoldsen — APPROVED** [review]
This looks good. The double encode + fetching seems a bit odd. Do you want me to take a look at that?

Besides these things, everything looks good. There is also some debug code that we have to clean up before we push.

File: mteb/models/repllama_models.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+
+class RepLLaMAWrapper:
+    def __init__(self, *args, **kwargs):
+        try:
+            from peft import PeftModel
+        except ImportError:
+            raise ImportError(
+                "To use the RepLLaMA based models `peft` is required. Please install it with `pip install peft`."
+            )
```
Comment (KennethEnevoldsen) [inline mteb/models/repllama_models.py:29]:
```suggestion
                "To use the RepLLaMA based models `peft` is required. Please install it with `pip install 'mteb[peft]'`."
            )
```

that way we can specify the version

**Comment by KennethEnevoldsen** [comment]
Tested out `mteb.get_model_meta("samaya-ai/promptriever-llama2-7b-v1")` can't seem to reproduce the error. 

Wanted to test out the double encode as well but can't access the model.

File: mteb/models/promptriever_models.py (hunk 1 above)
Comment (orionw) [reply mteb/models/promptriever_models.py]: Ah, great idea

**Comment by orionw** [comment]
> Tested out mteb.get_model_meta("samaya-ai/promptriever-llama2-7b-v1") can't seem to reproduce the error.

Ah sorry, it's public now.  And updated the hashes, great idea!

**Comment by KennethEnevoldsen** [comment]
@orionw I can't seem to get reproduce the error with:

```mteb.get_model_meta("samaya-ai/promptriever-llama2-7b-v1")```

It seems to work just fine on my end.

File: mteb/models/repllama_models.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+EncodeTypes = Literal["query", "passage"]
+
+
+class RepLLaMAWrapper:
+    def __init__(self, *args, **kwargs):
+        try:
+            from peft import PeftModel
+        except ImportError:
+            raise ImportError(
+                "To use the RepLLaMA based models `peft` is required. Please install it with `pip install 'mteb[peft]'`."
```
Comment (Samoed) [inline mteb/models/repllama_models.py:28]: I think that you should add peft to additional dependencies in pyproject

Comment (orionw) [reply mteb/models/repllama_models.py:28]: Ah great point, I thought it was already there because of LLM2Vec.  Added!

**Comment by orionw** [comment]
@KennethEnevoldsen thanks for checking, I must have something weird in my new local environment. Seems good to merge then, assuming the pyproject commit I just added doesn't break things.

**Pull request merged**